### PR TITLE
OSSL_STORE and APPS: minor improvements on credential loading and docs

### PR DESCRIPTION
**Pull request by DDvO**
* APPS `load_key_certs_crls()`: Correct the `expect` arg calculation for `OSSL_STORE_expect()`
* `OSSL_STORE_expect()`: Improve error handling and documentation
* `BIO_eof()` and `OSSL_STORE_eof()`: Correct doc and use of return value < 0 in `store_lib.c`
* `OSSL_DECODER_from_bio()`: Prevent spurious decoding error at EOF

##### Checklist

- [x] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

File: crypto/encode_decode/decoder_lib.c [hunk 1]
```diff
@@ -79,10 +79,11 @@ int OSSL_DECODER_from_bio(OSSL_DECODER_CTX *ctx, BIO *in)
         const char *input_structure
             = ctx->input_structure != NULL ? ctx->input_structure : "";
 
-        ERR_raise_data(ERR_LIB_OSSL_DECODER, ERR_R_UNSUPPORTED,
-                       "No supported for the data to decode.%s%s%s%s%s%s",
-                       spaces, input_type_label, input_type, comma,
-                       input_structure_label, input_structure);
+        if (BIO_eof(in) == 0)
+            ERR_raise_data(ERR_LIB_OSSL_DECODER, ERR_R_UNSUPPORTED,
+                           "Not supported for the data to decode.%s%s%s%s%s%s",
```
Comment (t8m) [inline crypto/encode_decode/decoder_lib.c:84]: Probably rather `No support for the data to decode`.

Could you also please explain why the `if (BIO_eof(in) == 0)` is added?

File: crypto/store/store_lib.c [hunk 2]
```diff
@@ -240,6 +240,11 @@ int OSSL_STORE_expect(OSSL_STORE_CTX *ctx, int expected_type)
 {
     int ret = 1;
 
+    if (ctx == NULL
+            || 0 > expected_type || expected_type > OSSL_STORE_INFO_CRL) {
```
Comment (t8m) [inline crypto/store/store_lib.c:244]: Please use `expected_type < 0`

File: crypto/store/store_lib.c [hunk 3]
```diff
@@ -362,7 +367,7 @@ OSSL_STORE_INFO *OSSL_STORE_load(OSSL_STORE_CTX *ctx)
 
     ctx->loading = 1;
  again:
-    if (OSSL_STORE_eof(ctx))
+    if (OSSL_STORE_eof(ctx) == 1)
```
Comment (t8m) [inline crypto/store/store_lib.c:370]: Why this change? The OSSL_STORE_eof is supposed to return 0 or 1.

File: crypto/store/store_lib.c [hunk 4]
```diff
@@ -388,7 +393,7 @@ OSSL_STORE_INFO *OSSL_STORE_load(OSSL_STORE_CTX *ctx)
                                              &load_data,
                                              ossl_pw_passphrase_callback_dec,
                                              &ctx->pwdata)) {
-                if (!OSSL_STORE_eof(ctx))
+                if (OSSL_STORE_eof(ctx) != 1)
```
Comment (t8m) [inline crypto/store/store_lib.c:396]: Same here, please revert.

File: doc/man3/OSSL_STORE_open.pod [hunk 5]
```diff
@@ -143,8 +143,8 @@ on error or when end of data is reached.
 Use OSSL_STORE_error() and OSSL_STORE_eof() to determine the meaning of a
 returned NULL.
 
-OSSL_STORE_eof() returns 1 if the end of data has been reached, otherwise
-0.
+OSSL_STORE_eof() returns 1 if the end of data has been reached, 0 if not,
```
Comment (t8m) [inline doc/man3/OSSL_STORE_open.pod:146]: It would be really bad to change the API this way. It was not expected to return <0 on error in 1.1.1.

File: doc/man3/BIO_ctrl.pod [hunk 6]
```diff
@@ -92,7 +92,7 @@ for success and -1 for failure.
 
 BIO_flush() returns 1 for success and 0 or -1 for failure.
 
-BIO_eof() returns 1 if EOF has been reached 0 otherwise.
+BIO_eof() returns 1 if EOF has been reached, 0 if not, or < 0 on error.
```
Comment (t8m) [inline doc/man3/BIO_ctrl.pod:95]: No! This is wrong, you cannot change public API like this.

File: doc/man3/OSSL_STORE_open.pod (hunk 5 above)
Comment (DDvO) [reply doc/man3/OSSL_STORE_open.pod:146]: This is not an API change, but a correction of the documentation!
Note that `BIO_eof()` and thus `OSSL_STORE_eof()` CAN return -1 or -2 on error.
Or shall solve the discrepancy the other way round, making sure that <0 is mapped to 0?

File: doc/man3/BIO_ctrl.pod (hunk 6 above)
Comment (DDvO) [reply doc/man3/BIO_ctrl.pod:95]: Same as above.

File: crypto/store/store_lib.c (hunk 3 above)
Comment (DDvO) [reply crypto/store/store_lib.c:370]: See below.

File: crypto/store/store_lib.c (hunk 4 above)
Comment (DDvO) [reply crypto/store/store_lib.c:396]: See below.

File: apps/lib/apps.c [hunk 7]
```diff
@@ -828,6 +828,7 @@ int load_crls(const char *uri, STACK_OF(X509_CRL) **crls,
     return ret;
 }
 
+#define SET_EXPECT(val) expect = expect < 0 ? val : expect == val ? val : 0;
```
Comment (richsalz) [inline apps/lib/apps.c:831]: a comment explaining this would be helpful, as would the redundant parens after the first colon.

File: doc/man3/OSSL_STORE_open.pod (hunk 5 above)
Comment (levitte) [reply doc/man3/OSSL_STORE_open.pod:146]: OSSL_STORE_eof() should never return < 0.  I hadn't thought of what BIO_eof() can return...

Here's my suggestion:

``` diff
diff --git a/providers/implementations/storemgmt/file_store.c b/providers/implementations/storemgmt/file_store.c
index 37f2fcee67..c300698b8c 100644
--- a/providers/implementations/storemgmt/file_store.c
+++ b/providers/implementations/storemgmt/file_store.c
@@ -705,12 +705,23 @@ static int file_eof(void *loaderctx)
     case IS_DIR:
         return ctx->_.dir.end_reached;
     case IS_FILE:
+        int ret;
+
         /*
-         * BIO_pending() checks any filter BIO.
-         * BIO_eof() checks the source BIO.
+         * BIO_pending() checks any filter BIO.  If anything is pending, or
+         * there was an error, this can't be EOF.
          */
-        return !BIO_pending(ctx->_.file.file)
-            && BIO_eof(ctx->_.file.file);
+        ret = BIO_pending(ctx->_.file.file);
+        if (ret != 0)
+            return 0;
+
+        /* BIO_eof() checks the source BIO. */
+        ret = BIO_eof(ctx->_.file.file);
+        if (ret <= 0)
+            return 0;
+
+        /* We have reached the end of input */
+        return 1;
     }
 
     /* ctx->type has an unexpected value */
```

Comment (t8m) [reply doc/man3/OSSL_STORE_open.pod:146]: This can happen only if the CTRL is unimplemented, isn't it so?

Comment (t8m) [reply doc/man3/OSSL_STORE_open.pod:146]: BIO_eof is not documented to return -1 or -2 either. So it should be fixed instead.

Comment (t8m) [reply doc/man3/OSSL_STORE_open.pod:146]: IMO it returns -1 or -2 only on _impossible_ conditions.

Comment (levitte) [reply doc/man3/OSSL_STORE_open.pod:146]: > So it [BIO_eof()] should be fixed instead.

I dunno about that, the BIO_eof() behavior is much older (you can trace if *far* back), and coded directly.  That the file scheme implementation of OSSL_STORE_eof() happened to propagate it was accidental.

So, while you might want to fix BIO_eof(), the patch I proposed is still not *wrong*.

File: doc/man3/BIO_ctrl.pod (hunk 6 above)
Comment (t8m) [reply doc/man3/BIO_ctrl.pod:95]: As I said above this can happen only on _impossible_ errors such as BIO_CTRL_EOF or bio_ctrl call not implemented. I'd then recommend to change the BIO_eof() code to return 1 in these cases instead of changing all the callers.

File: doc/man3/OSSL_STORE_open.pod (hunk 5 above)
Comment (DDvO) [reply doc/man3/OSSL_STORE_open.pod:146]: How about adapting in `bio.h.in`:
```
# define BIO_eof(b)              (int)BIO_ctrl(b,BIO_CTRL_EOF,0,NULL)
```
to
```
# define BIO_eof(b)              (BIO_ctrl(b,BIO_CTRL_EOF,0,NULL) == 1)
```
This way `file_eof()` and `OSSL_STORE_eof()` get fixed as well.

File: crypto/encode_decode/decoder_lib.c (hunk 1 above)
Comment (DDvO) [reply crypto/encode_decode/decoder_lib.c:84]: It's explained above and in the commit message:
```
OSSL_DECODER_from_bio(): Prevent spurious decoding error at EOF
```
but I can also add a comment in the code.

File: doc/man3/BIO_ctrl.pod (hunk 6 above)
Comment (DDvO) [reply doc/man3/BIO_ctrl.pod:95]: Yeah, let's do the correction at `BIO_eof()` level.

File: crypto/encode_decode/decoder_lib.c (hunk 1 above)
Comment (DDvO) [reply crypto/encode_decode/decoder_lib.c:84]: Done

File: crypto/store/store_lib.c (hunk 2 above)
Comment (DDvO) [reply crypto/store/store_lib.c]: Done

File: crypto/store/store_lib.c (hunk 3 above)
Comment (DDvO) [reply crypto/store/store_lib.c]: Reverted

File: doc/man3/OSSL_STORE_open.pod (hunk 5 above)
Comment (DDvO) [reply doc/man3/OSSL_STORE_open.pod]: I've just executed my proposal.

File: doc/man3/BIO_ctrl.pod (hunk 6 above)
Comment (DDvO) [reply doc/man3/BIO_ctrl.pod]: Done

File: apps/lib/apps.c (hunk 7 above)
Comment (DDvO) [reply apps/lib/apps.c]: Done

File: include/openssl/bio.h.in [hunk 8]
```diff
@@ -520,7 +520,7 @@ int BIO_read_filename(BIO *b, const char *name);
 # define BIO_dup_state(b,ret)    BIO_ctrl(b,BIO_CTRL_DUP,0,(char *)(ret))
 
 # define BIO_reset(b)            (int)BIO_ctrl(b,BIO_CTRL_RESET,0,NULL)
-# define BIO_eof(b)              (int)BIO_ctrl(b,BIO_CTRL_EOF,0,NULL)
+# define BIO_eof(b)              (BIO_ctrl(b,BIO_CTRL_EOF,0,NULL) == 1)
```
Comment (t8m) [inline include/openssl/bio.h.in:523]: IMO this is incorrect. It should be `BIO_ctrl(b,BIO_CTRL_EOF,0,NULL) != 0` instead. It is much more safe to assume error means EOF than the other way around as an error means that you most probably won't be able to read any data anyway. Also if the current callers just used if (BIO_eof(b)) your proposal will change the behavior.

Comment (DDvO) [reply include/openssl/bio.h.in]: Good point - done.
This is also more consistent with existing use of the eof() functions.

**Review by t8m — APPROVED** [review]
LGTM now

File: doc/man3/BIO_ctrl.pod [hunk 9]
```diff
@@ -92,7 +92,7 @@ for success and -1 for failure.
 
 BIO_flush() returns 1 for success and 0 or -1 for failure.
 
-BIO_eof() returns 1 if EOF has been reached 0 otherwise.
+BIO_eof() returns 1 if EOF has been reached or an error occurred, 0 otherwise.
```
Comment (levitte) [inline doc/man3/BIO_ctrl.pod:95]: This is really bad, 'cause now, there's no way to distinguish if EOF was reached (*reaching EOF is **not** an error!!!*) or an actual error occurred.

One way of solving this would be to add BIO_error(), which would enable the same sort of interface as feof() and ferror().

Comment (DDvO) [reply doc/man3/BIO_ctrl.pod:95]: I agree in principle, but according to @t8m the error condition does not really occur in practice (which I did not check) and at least so far users will have implicitly used `!= 0` anyway, relying on the hitherto (not adequately) documented behavior.

I suggest opening a separate issue/PR on adding `BIO_error()`.

File: include/openssl/bio.h.in [hunk 10]
```diff
@@ -520,7 +520,7 @@ int BIO_read_filename(BIO *b, const char *name);
 # define BIO_dup_state(b,ret)    BIO_ctrl(b,BIO_CTRL_DUP,0,(char *)(ret))
 
 # define BIO_reset(b)            (int)BIO_ctrl(b,BIO_CTRL_RESET,0,NULL)
-# define BIO_eof(b)              (int)BIO_ctrl(b,BIO_CTRL_EOF,0,NULL)
+# define BIO_eof(b)              (BIO_ctrl(b,BIO_CTRL_EOF,0,NULL) != 0)
```
Comment (levitte) [inline include/openssl/bio.h.in:523]:
``` suggestion
# define BIO_eof(b)              (BIO_ctrl(b,BIO_CTRL_EOF,0,NULL) > 0)
```

At the very least, it avoids making EOF and errors equivalent

File: doc/man3/BIO_ctrl.pod (hunk 9 above)
Comment (t8m) [reply doc/man3/BIO_ctrl.pod:95]: @levitte Please look at how the BIO_CTRL_EOF is implemented. In our BIOs a negative result can currently happen only if BIO_CTRL_EOF is unimplemented or there is no ctrl call at all in the BIO method. And all the callers IMO currently treat negative result as EOF anyway. So I do not think this is bad change at all. It just makes the current behavior more "official".

File: include/openssl/bio.h.in (hunk 10 above)
Comment (t8m) [reply include/openssl/bio.h.in:523]: No! All the current callers treat the negative return of BIO_eof() as if EOF was signaled and for a good reason. An EOF means to stop trying reading further. The current negative returns from BIO_eof() are unrecoverable, meaning no further call of BIO_eof() can succeed anyway. Returning 0 instead of -1 would signalize to the callers that it makes sense to try to read further and that would be wrong.

Comment (levitte) [reply include/openssl/bio.h.in:523]: It's quite obvious that changing things at this level is plain wrong.

-1

Comment (DDvO) [reply include/openssl/bio.h.in:523]: Let's hand out duel guns :D

**Comment by DDvO** [comment]
This PR is approved by @t8m, but then @levitte contradicted to the solution - how to handle this?

**Comment by t8m** [comment]
> This PR is approved by @t8m, but then @levitte contradicted to the solution - how to handle this?

IMO an acceptable solution would be to keep the BIO_eof implementation as is, and possibly document the negative error return cases.

OSSL_store_eof already works fine. I.E., it does NOT return negative values and it treats BIO_eof error as EOF, which is correct.

**Comment by DDvO** [comment]
> OSSL_store_eof already works fine. I.E., it does NOT return negative values and it treats BIO_eof error as EOF, which is correct.

This is what i originally did here.

> OSSL_store_eof already works fine. I.E., it does NOT return negative values and it treats BIO_eof error as EOF, which is correct.

How can you be sure that it cannot return < 0?
How about 
changing its last statement to `return ret != 0` (to be on the safe sice) 
and documenting that it returns 1 also on error?

**Review by t8m — APPROVED** [review]
LGTM now

**Comment by DDvO** [comment]
Merged - thanks @t8m
